cli/cmdlineparser.cpp

```diff
@@ -488,6 +490,17 @@ CmdLineParser::Result CmdLineParser::parseFromArgs(int argc, const char* const a
                 mSettings.checkLibrary = true;
             }
 
+            else if (std::strncmp(argv[i], "--check-version=", 16) == 0) {
```

Comment:
This needs to be an exclusive option.

Reply:
Then I don't think you understand the purpose. It's an additional parameter used in your normal cppcheck analysis command.

Reply:
Yeah. I do not understand the point of this at all - that's why I asked for the use case.

It seems like somebody who uses the CLI doesn't know how to or doesn't want to script and wants that logic added to the application.

Reply:
> It seems like somebody who uses the CLI doesn't know how to or doesn't want to script and wants that logic added to the application.

Yes. I think you understand the point then. We pin the uncrustify version in our runformat script.. we do it with some script commands but this option will make it possible to pin the cppcheck version without "complex" scripting.

It was an important point in the safety certification that we prevent that users run the wrong version by mistake.

Reply:
I hope you can be open to have a bit of extra options/code for safety. It will mean a lot for users that wants to check automotive/medical/etc projects.

This option will be something that I will require that certain users use, for safety reasons. I could in theory require that the users check the version in a script instead, but I wouldn't be able to document how to properly do that in every possible script language.

Some of the "safety-related" code is not wanted in non-safety projects then we can hide it behind the `--safety` option. I am fine with hiding any "safety-related" code behind that. For instance I don't know if you wanted to remove the checkers report from non-safety completely..

Reply:
>  This option is similar in spirit but will make it possible to pin the cppcheck version without "complex" scripting.

If you add `--version-short` which only prints the actual version there would be no "complex" scripting involved at all since it would be simple string comparison. Also without a short version it would just be a string compare in a script.

This also introduces issues depending on the integration since it will fail but you need to parse the command-line to get the error. So it will actually become "complex" naturally. Similar to the deprecations I want to introduce.

> It was an important point in the safety certification that we prevent that users run the wrong version by mistake.

How can a wrong version appear?

I would get this with something like `pypi` where arbitrary version would be pulled from remote. Or a hash/certificate if you communicate with a remote server. But in our case you need to actively install a new version on a system and you are not accessing the application remotely. And if you update the version you also need to update any users. So I wonder how this could get out-of-sync. I am still lacking the actual use case here.

It seems like during the certification it was missed that this is a local application.

And how does the user know what the actual version is? They would probably just look at the version of the binary and then specify that. Or leave out the parameter and then it works.

It seems to me like this should be something mandatory so without the parameter you are not allowed to run it at all. You could enable that through `cppcheck.cfg` (which would need to be read-only so it cannot be modified or removed) or via a compiler switch. In that case you would also require it to obfuscate the version so it cannot simply be looked up. So you need something unique instead like a hash/key which needs to change with each binary. That way it would actually start to make sense. But as that key needs to live in the binary you also need to encrypt that and adding encryption in any application opens multiple cans of worms you do not want it since that lead to other safety requirement and export restrictions....

That current approach would be sufficient for the certification seems quite laughable.

Reply:
I need to explain the use case.. I have a meeting now for ~ 30-60 minutes.. will explain after that

Reply:
> I hope you can be open to have a bit of extra options/code for safety.

The main issue is that if we introduce such things it will be very unlikely that we can do much changes to them afterwards. So if these things are being rolled out somewhere they need to be right. Otherwise you would need to deprecate things, have backwards compatibility, legacy code and a fragmentation in the integrations and it would also complicate support. That would just annoy everybody involved.

As there's lots of stuff we still need to tinker with years after it has been introduced because we have not been good at making the right decision (even if it seemed to be an informed decision). So that's why I asked about the use case to understand why it is needed and what is tried to being achieved.

Also there might be an additional requirement that patch versions do not need to match or that you do not want older versions allowed but newer ones. Then the logic starts to pile up and we would implement logic which is already widely available in scripting libraries. So just a throwaway option might suddenly lead to feature creep...

Reply:
> I need to explain the use case.. I have a meeting now for ~ 30-60 minutes.. will explain after that

Please add the use case to the ticket. I will be out soon as well.

Reply:
> How can a wrong version appear?

It's the same as with our version check in "runformat". How could a cppcheck developer install the wrong uncrustify version?

Maybe a developer will install latest cppcheck 2.13 on his computer. But the CI machine and everybody else runs Cppcheck 2.12.

Or the developer forgets to upgrade Cppcheck when the CI is upgraded.

A user can also for instance have both Cppcheck Premium and open source Cppcheck on his computer. If the cppcheck command in the script is "cppcheck file.cpp" then whatever cppcheck binary that is found in the PATH will be used to check the file.cpp.

> And how does the user know what the actual version is? They would probably just look at the version of the binary and then specify that. Or leave out the parameter and then it works.

It is not supposed to be changed by the user like that. I don't change the runformat script to match my uncrustify version when I want to run that.

The use case is that the user runs a common script that is used by everybody in the team. That script executes cppcheck on the code. It makes sense that everybody in the team uses the same script and that it's the same script that is used in CI.

Reply:
Sorry for the late reply. First I was not feeling well and then was busy for the whole extended weekend.

Okay, the missing information was that it will *only* be invoked by a script. I finally have the actual use case.

But I still think this in unnecessary since it is just comparing a string which is possible in any script language. And if you are not be able to do that...
And if you are running Cppcheck from a script only you will need to be slightly more complex things than that. And you probably also need to do something with the output from Cppcheck so you are basically already doing the same just from a different invocation.

Reply:
But that would theoretically introduce a "Time-of-check to time-of-use" race condition. And I totally forgot about the safety certification background.

I guess it is fine. I do not like it and I think it sets a bit of a precedent for allowing things/logic in which shouldn't.